File: Test Case Generation/deepseek/cpp/test_task_3.cpp
#include <gtest/gtest.h>
#include <vector>

// Function declaration (assuming it's in the global namespace)
bool below_zero(std::vector<int> operations);

// Test suite for below_zero function
class BelowZeroTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// Test: Empty vector - never below zero
TEST_F(BelowZeroTest, EmptyVectorReturnsFalse) {
    std::vector<int> operations;
    EXPECT_FALSE(below_zero(operations));
}

// Test: Single positive deposit
TEST_F(BelowZeroTest, SinglePositiveDepositReturnsFalse) {
    std::vector<int> operations = {100};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Single negative deposit (immediately below zero)
TEST_F(BelowZeroTest, SingleNegativeDepositReturnsTrue) {
    std::vector<int> operations = {-5};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Single zero deposit
TEST_F(BelowZeroTest, SingleZeroDepositReturnsFalse) {
    std::vector<int> operations = {0};
    EXPECT_FALSE(below_zero(operations));
}

// Test: All positive deposits
TEST_F(BelowZeroTest, AllPositiveDepositsReturnsFalse) {
    std::vector<int> operations = {1, 2, 3, 4, 5};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Balance never drops below zero
TEST_F(BelowZeroTest, BalanceNeverBelowZeroReturnsFalse) {
    std::vector<int> operations = {5, 3, -2, 4, -1};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Balance drops below zero exactly at the negative operation
TEST_F(BelowZeroTest, BalanceDropsBelowZeroAtNegativeOperationReturnsTrue) {
    std::vector<int> operations = {5, 3, -9};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Balance drops below zero after multiple operations
TEST_F(BelowZeroTest, BalanceDropsBelowZeroAfterMultipleOperationsReturnsTrue) {
    std::vector<int> operations = {1, 2, 3, -10};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Balance drops below zero then recovers but should still return true
TEST_F(BelowZeroTest, BalanceDropsBelowZeroThenRecoversReturnsTrue) {
    std::vector<int> operations = {1, -5, 10};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Balance drops below zero multiple times
TEST_F(BelowZeroTest, BalanceDropsBelowZeroMultipleTimesReturnsTrue) {
    std::vector<int> operations = {1, -2, 5, -10, 8};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Balance exactly zero after operations
TEST_F(BelowZeroTest, BalanceExactlyZeroAfterOperationsReturnsFalse) {
    std::vector<int> operations = {3, -2, 5, -6};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Balance exactly zero at intermediate point
TEST_F(BelowZeroTest, BalanceExactlyZeroAtIntermediatePointReturnsFalse) {
    std::vector<int> operations = {5, -5, 2};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Large positive values
TEST_F(BelowZeroTest, LargePositiveValuesReturnsFalse) {
    std::vector<int> operations = {1000000, 2000000, -1500000, 500000};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Large negative value causing immediate underflow
TEST_F(BelowZeroTest, LargeNegativeValueReturnsTrue) {
    std::vector<int> operations = {500000, -1000000};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Operation sequence that dips to negative, returns to positive, dips again
TEST_F(BelowZeroTest, MultipleNegativeDipsReturnsTrue) {
    std::vector<int> operations = {2, -3, 5, -8, 10, -12};
    EXPECT_TRUE(below_zero(operations));
}

// Test: Maximum integer values (assuming no overflow concerns)
TEST_F(BelowZeroTest, MaximumIntegerValues) {
    std::vector<int> operations = {2147483647, -2147483647};
    EXPECT_FALSE(below_zero(operations));
}

// Test: Minimum integer negative value
TEST_F(BelowZeroTest, MinimumIntegerNegativeValue) {
    std::vector<int> operations = {0, -2147483648};
    // This will underflow because adding negative max to zero will go negative
    EXPECT_TRUE(below_zero(operations));
}

File: Test Case Generation/deepseek/python/test_task_3.py
import pytest
from typing import List
from below_zero import below_zero  # assuming the function is in below_zero.py

def test_empty_list_returns_false():
    """Test with empty list of operations."""
    assert below_zero([]) is False

def test_single_positive_operation_returns_false():
    """Test with single positive deposit."""
    assert below_zero([5]) is False

def test_single_zero_operation_returns_false():
    """Test with single zero operation."""
    assert below_zero([0]) is False

def test_single_negative_operation_returns_true():
    """Test with single negative withdrawal that makes balance negative."""
    assert below_zero([-1]) is True

def test_multiple_positive_operations_returns_false():
    """Test with operations that keep balance positive or zero."""
    assert below_zero([1, 2, 3]) is False
    assert below_zero([0, 0, 0]) is False
    assert below_zero([5, 10, 15]) is False

def test_balance_never_negative_returns_false():
    """Test with mix of deposits and withdrawals but balance never negative."""
    assert below_zero([10, -5, 3, -2, 1]) is False
    assert below_zero([100, -50, -30, 20]) is False

def test_balance_becomes_negative_mid_sequence_returns_true():
    """Test where balance becomes negative during operations."""
    assert below_zero([1, -2, 3]) is True  # 1 -> -1 (negative)
    assert below_zero([5, -10, 2]) is True  # 5 -> -5 (negative)
    assert below_zero([1, 2, -4, 5]) is True  # 1,3,-1 (negative)
    assert below_zero([-5, 10]) is True  # -5 (immediately negative)

def test_balance_negative_at_last_operation_returns_true():
    """Test where balance becomes negative only at the final operation."""
    assert below_zero([3, 2, -6]) is True  # 3,5,-1 (negative at end)
    assert below_zero([10, 5, -20]) is True  # 10,15,-5 (negative at end)

def test_balance_zero_after_operations_returns_false():
    """Test where balance hits zero but never negative."""
    assert below_zero([5, -5, 2, -2]) is False
    assert below_zero([10, -5, -5]) is False

def test_large_numbers_handled_correctly():
    """Test with large integer values."""
    assert below_zero([1000000, -500000, -600000]) is True
    assert below_zero([1000000, -500000, -499999]) is False
    assert below_zero([-9223372036854775808]) is True  # min int

def test_balance_becomes_negative_then_positive_returns_true():
    """Test where balance becomes negative, then recovers - should still return True."""
    assert below_zero([5, -10, 20]) is True  # Goes to -5 at second operation
    assert below_zero([1, -2, 1, 5]) is True  # Goes to -1 at second operation

def test_consecutive_withdrawals_cause_negative():
    """Test with multiple withdrawals in a row causing negative balance."""
    assert below_zero([10, -3, -4, -5]) is True  # 10,7,3,-2 (negative)
    assert below_zero([20, -5, -10, -6]) is True  # 20,15,5,-1 (negative)

def test_balance_remains_positive_zero_never_negative():
    """Test various scenarios where balance never falls below zero."""
    assert below_zero([0, 0, 0, 0]) is False
    assert below_zero([1, -1, 2, -2, 3, -3]) is False
    assert below_zero([100, -50, -30, -20]) is False
    assert below_zero([5, 5, -2, -3, -5]) is False  # 5,10,8,5,0

def test_single_element_zero_returns_false():
    """Test with single zero operation."""
    assert below_zero([0]) is False

def test_first_operation_causes_negative():
    """Test where the very first operation makes balance negative."""
    assert below_zero([-10, 20, 30]) is True
    assert below_zero([-1, 100]) is True